File: RPC/RPC_04_04_05_2024/SolucionesOficiales/J_cake/Cake.java

// Problem: Cut the Cake
// Solution Author: Coach Travis

// Import
import java.util.*;

// The class containing the main method
public class Cake 
{
    // The scanner for reading from standard input
    static Scanner sc = new Scanner(System.in);

    // Method to get the difference of 2 2D vectors
    static int[] diff(int[] a, int[] b) 
    {
        return new int[]{a[0]-b[0],a[1]-b[1]};
    }

    // Method to get the cross product of 2 2D vectors
    static int cross(int[] a, int[] b) 
    {
        return a[0]*b[1] - a[1]*b[0];    
    }

    // Method to get the area of a triangle
    static int triangleArea(int[] a, int[] b, int[] c) 
    {
        int[] d = diff(b, a);
        int[] e = diff(c, a);
        return Math.abs(cross(d, e));
    }
    
    // Method to get the area of a polygon
    static int getArea(ArrayList<int[]>pts) {
        int area = 0;
        for (int i = 0; i < pts.size() - 1; i++) 
        {
            area += triangleArea(pts.get(0), pts.get(i), pts.get(i+1));
        }
        return area;
    }

    // The main method (entry point of the program)
    public static void main(String[] Args) 
    {
        // Read in the number of points
        int n = sc.nextInt();

        // Read in the points
        int[][] pts = new int[n][2];
        for (int i = 0; i < n; i++) 
        {
            pts[i][0] = sc.nextInt();
            pts[i][1] = sc.nextInt();
        }

        // Create variables for holding the answer
        long ans_area_one = 0;
        long ans_area_two = 1000000000;

        // Try all cut pairs
        for (int i = 0; i < n; i++) 
        {
            for (int j = i+2; j < n; j++) 
            {
                // Gen both sides of the cut
                ArrayList<int[]> pts2 = new ArrayList<>();
                for (int k = i; k <= j; k++)
                    pts2.add(pts[k]);
                ArrayList<int[]> pts3 = new ArrayList<>();
                for (int k = j; k != i + 1; k=(k+1)%n)
                    pts3.add(pts[k]);
        
                // Get their areas        
                int a = getArea(pts2);
                int b = getArea(pts3);
                
                // Update answer if it is better
                if (Math.abs(a-b) < Math.abs(ans_area_one-ans_area_two)) 
                {
                    ans_area_one = a;
                    ans_area_two = b;
                }
            }
        }

        // Print the answers
        long smallest = Math.min(ans_area_one, ans_area_two);
        long largest = Math.max(ans_area_one, ans_area_two);
        System.out.printf("%.01f%n", (largest - smallest) / 2.0);
    }
}

File: RPC/RPC_04_04_05_2024/SolucionesOficiales/J_cake/cake.cpp
// Arup Guha
// 8/26/2023
// Solution to UCF Locals 2023 Qualifying Round Problem: Cut the Cake

using namespace std;
#include <bits/stdc++.h>

int main() {

    // Read in the points.
    int n;
    cin >> n;
    vector<int> x(n);
    vector<int> y(n);
    for (int i=0; i<n; i++)
        cin >> x[i] >> y[i];

    // Get whole area (twice of it).
    int dWhole = 0;
    for (int i=0; i<n; i++)
        dWhole += (x[i]*y[(i+1)%n]-x[(i+1)%n]*y[i]);
    dWhole = abs(dWhole);

    // Answer is smaller than this.
    int dDiff = 2000000;

    // i is where we start the cut.
    for (int i=0; i<n; i++) {

        // j is where we end the cut.
        for (int j=i+2; j<n; j++) {

            // This is not a valid cut.
            if (i == 0 && j == n-1) continue;

            // Calculate the area of this polygon, from i to j and back.
            int thisD = 0;
            for (int k=i; k<j; k++)
                thisD += (x[k]*y[(k+1)%n]-x[(k+1)%n]*y[k]);
            thisD += (x[j]*y[i]-x[i]*y[j]);

            // This is double the area of this polygon.
            thisD = abs(thisD);

            // Get the rest.
            int other = dWhole - thisD;

            // Take the difference.
            int tmp = abs(thisD - other);

            // Update if necessary.
            dDiff = min(dDiff, tmp);
        }
    }

    // Ta da!
    printf("%.1f\n", dDiff/2.0);
    return 0;
}
